pkg/network/tracer/tracer_linux_test.go

```diff
@@ -1669,3 +1682,221 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+						c.Write(input)
+					}
+					c.Close()
+				})
+				require.NoError(t, server.Run(done))
+			},
+			want: network.ProtocolUnknown,
+		},
+		//{
+		//	//
+		//	name: "tcp client with sending encrypted data",
+		//	want: network.ProtocolUnknown,
+		//},
+		{
+			name:           "tcp client with sending HTTP request",
+			transportLayer: "tcp",
+			done:           make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				client := nethttp.Client{}
+				resp, err := client.Get("http://" + serverAddr + "/test")
+				require.NoError(t, err)
+				io.Copy(ioutil.Discard, resp.Body)
+				resp.Body.Close()
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				srv := &nethttp.Server{
+					Addr: serverAddr,
+					Handler: nethttp.HandlerFunc(func(w nethttp.ResponseWriter, req *nethttp.Request) {
+						io.Copy(ioutil.Discard, req.Body)
+						w.WriteHeader(200)
+					}),
+					ReadTimeout:  time.Second,
+					WriteTimeout: time.Second,
+				}
+				srv.SetKeepAlivesEnabled(false)
+				go func() {
+					_ = srv.ListenAndServe()
+				}()
+				go func() {
+					<-done
+					fmt.Println("shutting down")
+					srv.Shutdown(context.Background())
+				}()
+			},
+			want: network.ProtocolHTTP,
+		},
+		{
+			// Although it seems weird to test it, it is a real use-case.
+			// Consider a keep alive connection already established between a client and a server.
+			// Then the agent is being started, and able to capture the last response sent, and then the connection
+			// is terminated. Although we won't capture that HTTP traffic, it is still (as far as we know) a HTTP connection.
+			name:           "tcp client with sending HTTP response",
+			transportLayer: "tcp",
+			done:           make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				c, err := net.DialTimeout("tcp", serverAddr, time.Second)
+				require.NoError(t, err)
+				defer c.Close()
+				c.Write([]byte("HTTP/1.1 404 NOT FOUND\r\n"))
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				server := NewTCPServerOnAddress(serverAddr, func(c net.Conn) {
+					r := bufio.NewReader(c)
+					input, err := r.ReadBytes(byte('\n'))
+					if err == nil {
+						c.Write(input)
+					}
+					c.Close()
+				})
+				require.NoError(t, server.Run(done))
+			},
+			want: network.ProtocolHTTP,
+		},
+		//{
+		//	name: "tcp client with sending HTTP2 request",
+		//	want: network.ProtocolHTTP2,
+		//},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tr, err := NewTracer(cfg)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer tr.Stop()
+			defer close(tt.done)
+
+			initTracerState(t, tr)
+			port, err := GetFreePort()
+			require.NoError(t, err)
+			serverAddr := net.JoinHostPort(host, port)
+			fmt.Println("server", serverAddr)
+			tt.serverRun(t, serverAddr, tt.done)
+
+			time.Sleep(time.Second)
+			tt.clientRun(t, serverAddr)
+			time.Sleep(time.Second)
```

Comment:
does this sleeps are necessary?

Reply:
the first yes (although I reduced it to 500 milliseconds, and added a comment) - to let the server time to start running

the second - removed